libc/getpw.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libc-tests
+ *
+ * Testing getpwuid and getpwnam edge cases
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <pwd.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/stat.h>
+
+#include <unity_fixture.h>
+
+
+static unsigned int ispasswdfile;
+
+
+TEST_GROUP(getpwd);
+
+
+TEST_SETUP(getpwd)
+{
+}
+
+
+TEST_TEAR_DOWN(getpwd)
+{
+}
+
+
+/* Get root account details by name */
+TEST(getpwd, getpwnam_getroot)
+{
+	struct passwd *pw = getpwnam("root");
+
+	if (ispasswdfile) {
+		TEST_ASSERT_NOT_NULL(pw);
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_name);
+		/* no password hash content checking, only its nullability and size>0 */
+		TEST_ASSERT_NOT_NULL(pw->pw_passwd);
+		TEST_ASSERT_GREATER_THAN(0, strlen(pw->pw_passwd));
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_uid);
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_gid);
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_gecos);
+		TEST_ASSERT_EQUAL_STRING("/", pw->pw_dir);
+		TEST_ASSERT_EQUAL_STRING("/bin/sh", pw->pw_shell);
+	}
+	else {
+		TEST_ASSERT_NULL(pw);
+		TEST_IGNORE_MESSAGE("No /etc/passwd file!");
+	}
+}
+
+
+/* Get root account details by uid */
+TEST(getpwd, getpwuid_getroot)
+{
+	struct passwd *pw = getpwuid((uid_t)0);
+	if (ispasswdfile) {
+		TEST_ASSERT_NOT_NULL(pw);
+
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_name);
+		/* no password hash content checking, only its nullability and size>0 */
+		TEST_ASSERT_NOT_NULL(pw->pw_passwd);
+		TEST_ASSERT_GREATER_THAN(0, strlen(pw->pw_passwd));
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_uid);
+		TEST_ASSERT_EQUAL_INT(0, pw->pw_gid);
+		TEST_ASSERT_EQUAL_STRING("root", pw->pw_gecos);
+		TEST_ASSERT_EQUAL_STRING("/", pw->pw_dir);
+		TEST_ASSERT_EQUAL_STRING("/bin/sh", pw->pw_shell);
+	}
+	else {
+		TEST_ASSERT_NULL(pw);
+		TEST_IGNORE_MESSAGE("No /etc/passwd file!");
+	}
+}
+
+
+/* Try to reach for non-existing user data with getpwnam */
+TEST(getpwd, getpwnam_getnull)
+{
+	TEST_ASSERT_NULL(getpwnam("loremipsum"));
+	TEST_ASSERT_NULL(getpwnam(NULL));
+	/* TODO: add errno check */
+
+	if (!ispasswdfile)
+		TEST_IGNORE_MESSAGE("No /etc/passwd file!");
+}
+
+
+/* Try to reach for non-existing user data with getpwuid */
+TEST(getpwd, getpwuid_getnull)
+{
+	TEST_ASSERT_NULL(getpwuid((uid_t)-1));
+	TEST_ASSERT_NULL(getpwuid((uid_t)65535));
+	/* TODO: add errno check */
+
+	if (!ispasswdfile)
+		TEST_IGNORE_MESSAGE("No /etc/passwd file!");
+}
+
+
+/* Try to reach for root user data with no /etc/passwd file with getpwnam */
+TEST(getpwd, getpwnam_nopasswdfile)
+{
+	if (ispasswdfile)
+		rename("/etc/passwd", "/etc/passwd_del");
```

Comment:
note: this may fail on targets which will have readonly root partition. I don't know what should be done in that case (should we ignore the test then?). No need to change anything right now, just keep it in mind in the future.